src/query/graphite/graphite/timespec.go

```diff
@@ -108,9 +137,148 @@ func ParseTime(s string, now time.Time, absoluteOffset time.Duration) (time.Time
 		return time.Unix(n, 0).UTC(), nil
 	}
 
+	ref, offset := s, ""
+	if strings.Contains(s, "+") {
+		stringSlice := strings.Split(s, "+")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "+" + offset
+	} else if strings.Contains(s, "-") {
+		stringSlice := strings.Split(s, "-")
+		ref, offset = stringSlice[0], stringSlice[1]
+		offset = "-" + offset
+	}
+	parsedReference, err := ParseTimeReference(ref, now)
+	if err == nil {
+		parsedOffset, err := ParseOffset(offset)
+		if err == nil {
+			return parsedReference.Add(parsedOffset), nil
+		}
+	}
+
 	return now, err
 }
 
+func ParseTimeReference(ref string, now time.Time) (time.Time, error) {
+	var (
+		hour    = now.Hour()
+		minute  = now.Minute()
+		refDate time.Time
+	)
+
+	if ref == "" || ref == "now" {
+		return now, nil
+	}
+
+	// check if the time ref fits an absolute time format
+	for _, format := range formats {
+		t, err := time.Parse(format, ref)
+		if err == nil {
+			return time.Date(t.Year(), t.Month(), t.Day(), hour, minute, 0, 0, now.Location()), nil
+		}
+	}
+
+	rawRef := ref
+
+	// Time of Day Reference
+	i := strings.Index(ref, ":")
+	if 0 < i && i < 3 {
+		newHour, err := strconv.ParseInt(ref[:i], 10, 0)
+		if err != nil {
+			return time.Time{}, err
+		}
+		hour = int(newHour)
+		if len(ref) >= i+3 {
+			newMinute, err := strconv.ParseInt(ref[i+1:i+3], 10, 32)
+			if err != nil {
+				return time.Time{}, err
+			}
+			minute = int(newMinute)
+			ref = ref[i+3:]
+		}
+
+		if len(ref) >= 2 {
+			if ref[:2] == "am" {
+				ref = ref[2:]
+			} else if ref[:2] == "pm" {
+				hour = (hour + 12) % 24
```

Comment:
Is this valid for e.g. `23:00PM` or should that be a parse error

Reply:
Good call. I will throw an error and add a test case for this.